src/test/app/AMM_test.cpp

```diff
@@ -0,0 +1,6882 @@
+                    std::nullopt,
+                    ter);
+            };
+            for (int i = 0; i < 7; ++i)
+                vote(i, 10000000);
+            BEAST_EXPECT(ammAlice.expectTradingFee(175));
+            vote(7, 1000000, ter(tecAMM_FAILED_VOTE));
+        });
+    }
+
+    void
+    testFeeVote()
+    {
+        testcase("Fee Vote");
+        using namespace jtx;
+
+        // One vote sets fee to 1%.
+        testAMM([&](AMM& ammAlice, Env& env) {
+            ammAlice.vote({}, 1000);
+            BEAST_EXPECT(ammAlice.expectTradingFee(1000));
+        });
+
+        // Eight votes fill all voting slots, set fee 0.175%.
+        testAMM([&](AMM& ammAlice, Env& env) {
+            for (int i = 0; i < 7; ++i)
+            {
+                Account a(std::to_string(i));
+                fund(env, gw, {a}, {USD(10000)}, Fund::Acct);
+                ammAlice.deposit(a, 10000000);
+                ammAlice.vote(a, 50 * (i + 1));
+            }
+            BEAST_EXPECT(ammAlice.expectTradingFee(175));
+        });
+
+        // Eight votes fill all voting slots, set fee 0.175%.
+        // New vote, same account, sets fee 0.275%
+        testAMM([&](AMM& ammAlice, Env& env) {
+            auto vote = [&](Account const& a, int i) {
+                fund(env, gw, {a}, {USD(100000)}, Fund::Acct);
+                ammAlice.deposit(a, 10000000);
+                ammAlice.vote(a, 50 * (i + 1));
+            };
+            Account a("0");
+            vote(a, 0);
+            for (int i = 1; i < 7; ++i)
+            {
+                Account a(std::to_string(i));
+                vote(a, i);
+            }
+            BEAST_EXPECT(ammAlice.expectTradingFee(175));
+            ammAlice.vote(a, 450);
+            BEAST_EXPECT(ammAlice.expectTradingFee(225));
+        });
+
+        // Eight votes fill all voting slots, set fee 0.225%.
+        // New vote, new account, higher vote weight, set higher fee 0.244%
+        testAMM([&](AMM& ammAlice, Env& env) {
+            auto vote = [&](int i, std::uint32_t tokens) {
+                Account a(std::to_string(i));
+                fund(env, gw, {a}, {USD(100000)}, Fund::Acct);
+                ammAlice.deposit(a, tokens);
+                ammAlice.vote(a, 50 * (i + 1));
+            };
+            for (int i = 0; i < 7; ++i)
+                vote(i, 10000000);
+            BEAST_EXPECT(ammAlice.expectTradingFee(175));
+            vote(7, 20000000);
+            BEAST_EXPECT(ammAlice.expectTradingFee(244));
+        });
+
+        // Eight votes fill all voting slots, set fee 0.219%.
+        // New vote, new account, higher vote weight, set smaller fee 0.206%
+        testAMM([&](AMM& ammAlice, Env& env) {
+            auto vote = [&](int i, std::uint32_t tokens) {
+                Account a(std::to_string(i));
+                fund(env, gw, {a}, {USD(100000)}, Fund::Acct);
+                ammAlice.deposit(a, tokens);
+                ammAlice.vote(a, 50 * (i + 1));
+            };
```

Comment:
The `vote` lambdas are so similar in this function that I figured it would be worth refactoring to use one common function. That's at https://github.com/ximinez/rippled/commit/800fa54d1dbacc085ba6b0038c4c0e2b6653a833. (As I may have mentioned before, that branch isn't stable overall, but the commit should be good to cherry pick.)